File: 백준/Silver/1388. 바닥 장식/바닥 장식.java
import java.io.*;
import java.util.*;


public class Main {
    static StringTokenizer st;
    static boolean[][] visited;
    static char[][] arr;
    static byte N,M;
    static void dfs1(int x, int y) {
        visited[x][y] = true;
        if (y + 1 < M && arr[x][y + 1] == '-') {
            dfs1(x, y + 1);
        }
        return;
    }
    static void dfs2(int x, int y) {
        visited[x][y] = true;
        if (x + 1 < N && arr[x+1][y] == '|') {
            dfs2(x+1, y);
        }
        return;
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        st = new StringTokenizer(br.readLine());
        N = Byte.parseByte(st.nextToken());
        M = Byte.parseByte(st.nextToken());
        arr = new char[N][M];
        visited = new boolean[N][M];
        int ans = 0;
        for (int i = 0; i < N; i++) {
            arr[i] = br.readLine().toCharArray();
        }
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (visited[i][j]) {
                    continue;
                }
                ans += 1;
                if (arr[i][j] == '-') {
                    dfs1(i, j);
                } else {
                    dfs2(i,j);
                }

            }
        }
        bw.write(Integer.toString(ans));
        bw.close();
    }

}

File: 백준/Silver/1388. 바닥 장식/바닥 장식.py
def dfs1(x,y):
    visited[x][y]= True
    if y+1<M and arr[x][y+1]=="-":
        dfs1(x,y+1)
    return
def dfs2(x,y):
    visited[x][y]= True
    if x+1<N and arr[x+1][y]=="|":
        dfs2(x+1,y)
    return

N,M = map(int,input().split())
arr = [input() for _ in range(N)]
visited = [[False]*M for _ in range(N)]
ans = 0
for i in range(N):
    for j in range(M):
        if visited[i][j]:
            continue
        ans+=1
        if arr[i][j] =="-":
            dfs1(i,j)
        else:
            dfs2(i,j)
print(ans)